Update of Jameson's "common parametric method patterns"

mauro3 opened the pull request:
This updates https://github.com/JuliaLang/julia/pull/19306, to new markdown and fixes some outdated stuff.

Commit:
add a list of common parametric method patterns to the manual

this section will hopefully be helpful to new users both as motivational examples
demonstrating the usefulness of the ideas behind dispatch

as well as showing design patterns for more advanced users to refer to
when developing generic library code

Updated by mauro3 to convert .rst -> .md

[ci skip]

mauro3 committed:
Some extra updates to the manual section of common parametric method patterns

- linked references
- where syntax
- extra bits on `eltype` section
- clarification on traits section

mauro3 commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{A}) where A<:AbstractArray{T} where T = T
````
Is this the correct way these days?

tkelman replied:
a few `{}` would be good, we have tried to avoid one-liners that look like `... where T = T`

mauro3 replied:
Like so: `eltype(::Type{A}) where A<:AbstractArray{T} where {T} = T`?

fredrikekre replied:
I don't think there are many short function definitions with more than one `where` so not sure how it is usually done then. In this case the following works and is quite nice:
```julia
eltype(::Type{A}) where {T,A<:AbstractArray{T}} = T
```

mauro3 replied:
@fredrikekre I find that syntax a bit confusing.  In fact it seems a bit reminiscent of the oldern days (i.e. recently) when we had `eltype{T,A<:AbstractArray{T}}(::Type{A}) = T`.  So somehow it doesn't mix so well with the `where`.  Do you know what the advantage is to have both types of syntax?

This works too: `eltype(::Type{<:AbstractArray{T}}) where {T} = T`, maybe nicest?

fredrikekre replied:
FWIW `eltype{T,A<:AbstractArray{T}}(::Type{A}) = T` does not work pre 0.6.

> ... maybe nicest?

No preference

vtjnash replied:
> eltype(::Type{<:AbstractArray{T}}) where {T} = T

I think is now the preferred style, as it is most concise. However, there's a minor change to make, to handle the case of calling `eltype(Array)`:

```julia
- on v0.7:
eltype(::Type{<:AbstractArray{T}}) where {T} = @isdefined(T) ? T : Any

- on v0.6:
eltype(::Type{<:AbstractArray{T}}) where {T} = T
eltype(::Type{<:AbstractArray}) = Any
```

tkelman commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+eltype{A<:AbstractArray}(::Type{A}) = eltype(supertype(A))
+```
+
+One common mistake is to try and get the element-type by using introspection::
````
now that these are markdown, shouldn't double the colons

alyst commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+abstract AbstractArray{T, N}
+eltype(::Type{A}) where A<:AbstractArray{T} where T = T
+```
+using, so-called, triangular dispatch.
````
AFAIK commas are not required here

vtjnash replied:
I think it would be clearer to specify this as a definition:

This pattern, where one type parameter is dependent on another, is known as "triangular dispatch".

tkelman commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+For instance, you might have some sort of abstract array with an arbitrary element type
+and want to write the computation on it with a specific element type.
+
+Another implication of the common mistake discussed above mentioned above,
```
either "discussed above" or "mentioned above," one of them is redundant (and I don't think there should be a comma here)

alyst replied:
It's not obvious for me how the introspection mistake (if it's meant here) is related to the type transform method issue.

mauro3 replied:
Yep, I don't see it either & I deleted it.  But if important, maybe @vtjnash can clarify.

alyst commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+```
+using, so-called, triangular dispatch.
+
+Another way, which used to be only correct way before the advent of
````
"the only"; or does "way" belong to "before"?

tkelman commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+since those both operate in terms of type traits.
+When iterating over a matrix, such as in the implementation of `map`,
+one important question is what order to use to traverse the data.
+When `AbstractArray` subtypes implement the `Base.IndexStype` trait,
```
IndexStyle

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+
+The subtypes of `AbstractArray` typically implement two methods for expressing this:
+A method to convert the input to a subtype of a specific `AbstractArray{T, N}` abstract type;
+and a method to make a new uninitialized matrix with a specific element type.
```
matrix -> array

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+and want to write the computation on it with a specific element type.
+
+Another implication of the common mistake discussed above mentioned above,
+is that we must implement a method for each type that describes how to compute this type transform.
```
"... is that for each _<"element  type" or "`AbstractArray` subtype"?>_ we must implement a method that ..."?

mauro3 replied:
I think it is both, so I insterted `AbstractArray{T}` there.

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+### Iterated dispatch
+
+In order to dispatch a multi-level parametric argument list,
+often it is best to separate each level of dispatch into separate functions.
```
"distinct functions"?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+### Trait-based dispatch
+
+A natural extension to the iterated dispatch above is to add a layer to
+method selection that identifies classes of types.
```
"groups of types"? "class" could be confused as another type-system term

mauro3 replied:
Now:
"A natural extension to the iterated dispatch above is to add a layer to
method selection that allows to dispatch on sets of types which are disjoint from
the sets defined by the type hierarchy."
Clearer?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+method selection that identifies classes of types.
+We could do this by writing out a `Union` of the types with similar characteristics.
+But then this list would not be extensible.
+However, by expressing the property as a "trait",
```
_by defining "traits"_? "trait" is synonymous to a "property", and the latter has not been introduced before

mauro3 replied:
Ok, I re-worded this section slightly: can you give it another read?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+We could do this by writing out a `Union` of the types with similar characteristics.
+But then this list would not be extensible.
+However, by expressing the property as a "trait",
+it is possible to express type behaviors in the abstract
```
~~in the abstract~~

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+it is possible to express type behaviors in the abstract
+in a way that is flexible to new additions, but which has no performance impact.
+
+A trait is implemented by defining a pure generic function which returns a
```
..._is implemented as a pure generic function that_...?

alyst commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+map(::Base.IndexLinear, a::AbstractArray, b::AbstractArray) = # linear-fast implementation
+```
+
+This trait-based mechanism is also present in the [`promote`](@ref) mechanism used by `+`.
````
..._trait-based **approach**_...?
..._`promote` mechanism **employed** by `+`_...?

alyst commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+```
+
+This trait-based mechanism is also present in the [`promote`](@ref) mechanism used by `+`.
+That mechanism uses the types to compute the optimal common type for computing the operation.
````
_It uses the types of arguments to deduce the optimal common type for the result._?

gasagna commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+
+    # this is wrong, since depending on the return value
+    # of type-inference is very brittle (as well as not being optimizable):
+    # R = return_types(ap, (eltype(a), eltype(b)))
```
shouldn't this be `op`?

tkelman commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+When `AbstractArray` subtypes implement the `Base.IndexStype` trait,
+other functions such as `map` can dispatch on this information to pick
+the best algorithm (see [Abstract Array Interface](@ref man-interface-array)).
+This means that each subtypes does not need to implement a custom version of `map`,
```
each subtype

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+We could construct such a set by writing out a `Union` of the types in question,
+but then this set would not be extensible.
+However, by expressing such a set as a "Holy-trait",
+it is possible to allow new type additions to the set.
```
_to add new types to the set without redefining any existing types_?

mauro3 replied:
The point is if you "redefine" a Union, it will not be the same anymore.  Thus a function you wrote earlier on `f(::FirstUnion)` will still dispatch on that Union.

alyst replied:
Sure, every redefinition requires recompilation. That's one issue, but it's somewhat manageable. Actually, union-based dispatch should be ok for internal methods and is used in a number of places in `Base`.
The more serious issue that I tried to highlight is that union is "centralized", and e.g. there is no way for packages to extend the union defined in Base.
But I agree that my version is not explicit enough and still bears "run-time" vs "compile-time" confusion.

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+it is possible to allow new type additions to the set.
+
+Such a trait is implemented by defining a generic function which
+computes a different singleton value for each trait-set to which the
```
_different_ -> _distinct_?
is _singleton_ critical here? It adds complexion

mauro3 replied:
Yep.

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+
+Such a trait is implemented by defining a generic function which
+computes a different singleton value for each trait-set to which the
+function arguments may belong to.  If this function is pure there is
```
Maybe mention _singleton_ together with the function pureness?

mauro3 replied:
Ok

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+no impact on performance compared to normal dispatch.
+
+The example above glossed over the implementation details of [`map`](@ref) and [`promote`](@ref),
+since those both operate in terms of type traits.
```
_while the standard library implementation uses traits to ensure optimal performance_?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+function arguments may belong to.  If this function is pure there is
+no impact on performance compared to normal dispatch.
+
+The example above glossed over the implementation details of [`map`](@ref) and [`promote`](@ref),
```
_The matrix `+`/sum example above_

_glossed over_ -> _simplified_?
I admit I had to look into the dictionary for the meaning.

mauro3 replied:
No, I think that is fine and it made you learn a new word!

alyst commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+map(::Base.IndexLinear, a::AbstractArray, b::AbstractArray) = # linear-fast implementation
+```
+
+This trait-based approach is also present in the [`promote`](@ref) mechanism employed by `+`.
````
_scalar `+`_?
Otherwise it's confusing why speaking about `promote` again

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+
+For implementing primitive operations, such as addition,
+we use the [`promote_type`](@ref) function to compute the desired output type.
+(As before, we saw this at work in the `promote` call in the call to `+`).
```
_when calling `promote` from `+`_?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+type for a more complex sequence of operations.
+This is often performed by the following steps:
+
+1. Write a small function `op` that expresses the set of operations performed by the kernel of the algorithm.
```
_Define a small helper function `op` that exposes the numerical operations performed by the algorithm_?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+
+1. Write a small function `op` that expresses the set of operations performed by the kernel of the algorithm.
+2. Compute the element type `R` of the result matrix as `promote_op(op, argument_types...)`,
+   where `argument_types` is computed from `eltype` applied to each input array.
```
_`argument_types` are the element types of each input array_?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+1. Write a small function `op` that expresses the set of operations performed by the kernel of the algorithm.
+2. Compute the element type `R` of the result matrix as `promote_op(op, argument_types...)`,
+   where `argument_types` is computed from `eltype` applied to each input array.
+3. Build the output matrix as `similar(R, dims)`, where `dims` is the desired dimensions of the output array.
```
_are the dimensions_?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+plus a table of preferred pair-wise promotion rules.
+
+
+### Output-type Computation
```
lower case _c_

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+
+### Separate convert and kernel logic
+
+One way to significantly cut down on compile-times and testing complexity is to isolate
```
_cut down on ..._ -> _reduce compilation time and testing complexity_?

alyst commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+end
+```
+
+### Separate convert and kernel logic
````
"kernel" might be a bit slangy, and "logic" too; maybe _Separation of type conversion from computation_?
Alternatively "logic" -> "code"

mauro3 replied:
No, I think this is fine.

alyst replied:
Is "separate" a verb or an adjective here? If it's a verb, it doesn't align well with the style of the other section headings.

mauro3 replied:
Well, in that case, it's obviously an adjective 😄.  That is how I read it too.

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+### Separate convert and kernel logic
+
+One way to significantly cut down on compile-times and testing complexity is to isolate
+the logic for converting to the desired type and the computation.
```
_the code for converting to the desired type and the code for computation into separate methods_?

alyst commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+One way to significantly cut down on compile-times and testing complexity is to isolate
+the logic for converting to the desired type and the computation.
+This lets the compiler specialize and inline the conversion logic independent
+from the rest of the body of the larger kernel.
```
_This avoids recompilation of the larger computation part for each new combination of the input parameter types._?

mauro3 replied:
I think "kernel" is generic enough.  Also, once they get confused about this, I'd hope they have the means to de-confuse themselves.

mauro3 commented on doc/src/manual/methods.md:
```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+the sets defined by the type hierarchy.
+We could construct such a set by writing out a `Union` of the types in question,
+but then this set would not be extensible.
+However, by expressing such a set as a "Holy-trait",
```
I added the "Holy" to make it clearer that this is a pattern which allows to create a trait-like thing, as opposed to a language feature, which most doc-reading people (I would have thought) would it interpret as.  Maybe I could write "Such a set can be expressed with the following pattern..."

alyst replied:
Or _An alternative is to specify/describe sets of types using "Holy-traits" technique/code pattern (link) as it makes it possible to extend such sets without redefining any existing types_

vtjnash commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+eltype{T}(::Type{AbstractArray{T}}) = T
+eltype{T, N}(::Type{AbstractArray{T, N}}) = T
+eltype{A<:AbstractArray}(::Type{A}) = eltype(supertype(A))
+```
````
The proper way to write this in the new type (v0.6) system (with examples of which types would trigger each) is:
```julia
eltype(::Type{AbstractArray{T, N} where {T<:S, N<:M}}) where {M, S} = Any
  # eltype(Array) = Any
  # eltype(Array{T} where T <: Integer) = Any
eltype(::Type{AbstractArray{T, N} where {T<:S}}) where {N, S} = Any
  # eltype(Array{T, 1} where T <: Integer) = Any
eltype(::Type{AbstractArray{T, N} where {N<:M}}) where {M, T} = T
  # eltype(Array{Integer}) = Integer
  # eltype(Array{Integer, N} where N<:Integer) = Integer
eltype(::Type{AbstractArray{T, N}}) where {T, N} = T
  # eltype(Array{Integer, 3}) = Integer
eltype(::Type{A}) where {A <: AbstractArray} = eltype(supertype(A))
```

mauro3 replied:
This seems pretty complicated! More so than the old version, so we should either keep the old one or my update below, if it passes scrutiny.

I tried to update "my" version and this passes your tests:
```julia
f(::Type{<:AbstractArray{T}}) where {T} = T
f(::Type{<:AbstractArray}) = Any

@test f(Array{Integer, 3}) == Integer
@test f(Array{Integer}) == Integer
@test f(Array{Integer, N} where N<:Integer) == Integer
@test f(Array{T, 1} where T <: Integer) == Any
@test f(Array) == Any
@test f(Array{T} where T <: Integer) == Any
@test f(DenseArray{Integer}) == Integer
@test f(DenseArray{<:Integer}) == Any
```
The second methods seems to catch those `Any` cases.  Anything this misses?

On a more philosophical note, why not return `Integer` for `DenseArray{<:Integer}`? As all types which are subtypes of `DenseArray{<:Integer}` will have eltype `<:Integer`.

vtjnash replied:
Yes. I just wanted to show both in the manual so that people could make an informed choice :)

Also, I did indeed miss a class of cases:
```julia
julia> f(Array{Array{T,1}, 1} where T)
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] f(::Type{Array{Array{T,1},1} where T}) at ./REPL[1]:1
```
This is the case that shows why it needs `@isdefined(T) ? T : Any`.

(this also triggers what appears to be a type-system bug in "my" old version for which I'll file a separate issue)

vtjnash replied:
It doesn't return `Integer` because the `eltype` is a definite subtype of `Integer`, but it is not `Integer`.

mauro3 replied:
Ok, so I looked into this a bit more.  That error you report also happens to `eltype`:
```
julia> eltype(Array{Array{T, 1}, 1} where T)                                                                                                          
ERROR: UndefVarError: E not defined                                                                                                                   
Stacktrace:                                                                                                                                           
 [1] eltype(::Type{Array{Array{T,1},1} where T}) at ./abstractarray.jl:106                                                                            
```
but this is a bug in Julia, right? Or how can it be that the method matches but `T` is not defined:
```
julia> ff(::Type{<:AbstractArray{T}}) where {T} = T                                                                                                   
ff (generic function with 1 method)                                                                                                                   

julia> ff(Array{Array{T, 1}, 1} where T)                                                                                                              
ERROR: UndefVarError: T not defined                                                                                                                   
Stacktrace:                                                                                                                                           
 [1] ff(::Type{Array{Array{T,1},1} where T}) at ./REPL[3]:1                                                                                           
```
adding `@isdefined(T) ? T : Any` to `ff` does indeed fix it.

Anyway, as your methods and my two (and `eltype` itself) seem to be afflicted by this, I still think we should include the simpler version.

Aside, note that your methods do not work on 0.6 at all (presumably because of 0.6 bugs).  Here the gist for running all tests:
https://gist.github.com/mauro3/856423b2f746de67639314d68bec5e50

mauro3 replied:
> It doesn't return `Integer` because the `eltype` is a definite subtype of `Integer`, but it is not `Integer`.

There is also this case:
```
julia> eltype(Array{T, 1} where T<:Int)                                                                                                               
Any                                                                                                                                                   
```
where the eltype must be `Int`.  This could be fixed reasonably easy if `eltype(Array{T, 1} where T<:Integer)` would return `Integer`.

mauro3 replied:
@vtjnash any more thoughts on this and the results in the gist?

vtjnash replied:
My version runs into bug https://github.com/JuliaLang/julia/issues/23327. I think we should include all of the versions, with the comment that they are all equivalent ways of computing the same value. And leave it as an exercise for the reader to understand why.

mauro3 replied:
Ok, included.  See line 550: https://github.com/JuliaLang/julia/pull/23245/files#diff-619ea7b9ee7c2fa87a3720c1361ba46cR550

vtjnash commented on doc/src/manual/methods.md:
````diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+One common mistake is to try and get the element-type by using introspection:
+
+```julia
+eltype_wrong{A<:AbstractArray}(::Type{A}) = A.parameters[1]
````
update to use `where` syntax

vtjnash commented:
Thanks for taking the initiative on this! I haven't quite followed all of the discussion, so please ping me directly if there's any specific questions.

mauro3 commented on doc/src/manual/methods.md:
```diff
@@ -567,15 +567,13 @@ When building generic code, there is often a need for constructing a similar
 There is no general transform of one subtype into another subtype with a different parameter.
 (Quick review: do you see why this is?)
 
 The subtypes of `AbstractArray` typically implement two methods for expressing this:
```
I think so. Now `AbstractArray{<:T,N}`, which should make this clearer.

mauro3 commented:
Ok, did the updates.  Did a ci-skip on the last commit as I thought to save some cycles.  Will rebase once all good and we can ci again then.

alyst commented on doc/src/manual/methods.md:
```diff
@@ -618,43 +622,47 @@ This dispatching branching can be observed, for example, in the logic to sum two
 the best algorithm (see [Abstract Array Interface](@ref man-interface-array)).
 This means that each subtype does not need to implement a custom version of `map`,
 since the generic definitions + trait classes will enable the system to select the fastest version.
+Here a toy implementation of `map` illustrating the trait-based dispatch:
```
_Here's_

StefanKarpinski commented:
Can we make a last push on this and finally get it merged? Doesn't have to be perfect, just good enough!

mauro3 commented:
Yes, I can guarantee that it will not be perfect ;-)  I'll try to wrap it up today, latest tomorrow.

mauro3 commented:
Ok, I addressed most suggestions in the PR and this is ready to be merged as far as I'm concerned, thanks to all! 

One slight blemish is that currently the `eltype` example does not work for inputs as filed in issue #23327.  But I would think it is fine to merge this now, as no code in the manual actually fails.

StefanKarpinski commented:
Yes, let's definitely merge now (as soon as CI passes) and then fix any individual issues. Thanks so much for reviving this and seeing it through!

mauro3 commented:
There was one test failure https://travis-ci.org/JuliaLang/julia/jobs/267486400 in the `file`-tests, is this related?

StefanKarpinski commented:
Test failure is unrelated and ongoing.

mauro3 commented:
Note, if this is back-ported (I don't think it is necessary), then the lines https://github.com/JuliaLang/julia/blame/b64e005cdbcd0aa80e486af161cd8bf82845305c/doc/src/manual/methods.md#L550-L559 should be deleted, as this does not work in 0.6.

mauro3 commented:
@vtjnash, the fix of #23327 did fix the triangular dispatch `eltype` example (see our discussion above https://github.com/JuliaLang/julia/pull/23245#discussion_r133213159) but not your suggestion as it is currently in the docs:
```julia
eltype_(::Type{AbstractArray{T, N} where {T<:S, N<:M}}) where {M, S} = Any
eltype_(::Type{AbstractArray{T, N} where {T<:S}}) where {N, S} = Any
eltype_(::Type{AbstractArray{T, N} where {N<:M}}) where {M, T} = T
eltype_(::Type{AbstractArray{T, N}}) where {T, N} = T
eltype_(::Type{A}) where {A <: AbstractArray} = eltype_(supertype(A))

# These all throw
# ERROR: MethodError: no method matching eltype_(::Type{Any})    
eltype_(Array{Array{T, 1}, 1} where T) == Any
eltype_(Array{Array{T, N}, 1} where {N,T}) == Any
eltype_(Array{Array{T, N}, NN} where {NN,N,T}) == Any
```
Is that expected?

vtjnash commented:
Yeah, I suppose it is (I had this same discussion with Jeff on the Slack#types channel, since it's one of the changes his PR made). I guess this way of writing the method is simply just wrong, and the only correct way is to use:
```julia
eltype_(::Type{<:AbstractArray{T}}) where {T} = T
eltype_(::Type{<:AbstractArray}) = Any
```
